pingcap/tidb
# ddl: fix update with point get will try to set default value for the nil of not-null column

Signed-off-by: AilinKid <314806019@qq.com>



### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/22395

### What is changed and how it works?

What's Changed:
**add the origin default value when generate changing column(temporary column)**

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Manual test (add detailed scripts or steps below)

Here is a minimum reproduce script, the fixed PR should pass it.
```
package main

import (
	"database/sql"
	_ "github.com/go-sql-driver/mysql"
	"go.uber.org/zap"
	"math/rand"
	"strings"
)

func main() {
	logger, err := zap.NewProduction()
	if err != nil {
		panic(err)
	}

	db, _ := sql.Open("mysql", "root@tcp(127.0.0.1:4000)/test")
	_, _ = db.Exec("set global tidb_enable_change_column_type = 1")
	_, _ = db.Exec("drop table if exists t")
	_, _ = db.Exec("create table t (id int primary key, v int not null)")
	_, _ = db.Exec("insert into t values (0, 0), (1, 1), (2, 2)")
	_ = db.Close()

	for i := 0; i < 32; i++ {
		go func() {
			db, _ := sql.Open("mysql", "root@tcp(127.0.0.1:4000)/test")
			_, _ = db.Exec("set @@tidb_txn_mode = 'optimistic'")
			for {
				txn, err := db.Begin()
				if err != nil {
					panic(err)
				}
				id := rand.Intn(3)
				if _, err = txn.Exec("update t set v = ? where id = ?", rand.Intn(3), id); err != nil {
					panic(err)
				}
				if err = txn.Commit(); err != nil {
					if !strings.Contains(err.Error(), "8028") && !strings.Contains(err.Error(), "9007") {
						logger.Info("commit failure", zap.Error(err))
					}
				}
			}
		}()
	}

	db, _ = sql.Open("mysql", "root@tcp(127.0.0.1:4000)/test")
	for {
		if _, err = db.Exec("alter table t modify column v bigint not null"); err != nil {
			panic(err)
		}
		if _, err = db.Exec("alter table t modify column v int not null"); err != nil {
			panic(err)
		}
	}
}
```

### Release note 

- executor: fix update with point get will try to set default value for the nil of not-null column.
